builtin/logical/database/path_roles.go

```diff
@@ -2,263 +2,606 @@ package database
-		}
+func (b *databaseBackend) pathRoleExistenceCheck(ctx context.Context, req *logical.Request, data *framework.FieldData) (bool, error) {
+	role, err := b.Role(ctx, req.Storage, data.Get("name").(string))
+	if err != nil {
+		return false, err
+	}
+	return role != nil, nil
+}
 
-		return role != nil, nil
+func (b *databaseBackend) pathStaticRoleExistenceCheck(ctx context.Context, req *logical.Request, data *framework.FieldData) (bool, error) {
+	role, err := b.StaticRole(ctx, req.Storage, data.Get("name").(string))
+	if err != nil {
+		return false, err
 	}
+	return role != nil, nil
 }
 
-func (b *databaseBackend) pathRoleDelete() framework.OperationFunc {
-	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
-		err := req.Storage.Delete(ctx, "role/"+data.Get("name").(string))
-		if err != nil {
-			return nil, err
-		}
+func (b *databaseBackend) pathRoleDelete(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	err := req.Storage.Delete(ctx, databaseRolePath+data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
 
+func (b *databaseBackend) pathStaticRoleDelete(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+	// if this role is a static account, we need to revoke the user from the
+	// database
+	role, err := b.StaticRole(ctx, req.Storage, name)
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
 		return nil, nil
 	}
-}
 
-func (b *databaseBackend) pathRoleRead() framework.OperationFunc {
-	return func(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-		role, err := b.Role(ctx, req.Storage, d.Get("name").(string))
+	// clean up the static useraccount, if it exists
+	if role.StaticAccount != nil {
+		db, err := b.GetConnection(ctx, req.Storage, role.DBName)
 		if err != nil {
 			return nil, err
 		}
-		if role == nil {
-			return nil, nil
-		}
 
-		data := map[string]interface{}{
-			"db_name":               role.DBName,
-			"creation_statements":   role.Statements.Creation,
-			"revocation_statements": role.Statements.Revocation,
-			"rollback_statements":   role.Statements.Rollback,
-			"renew_statements":      role.Statements.Renewal,
-			"default_ttl":           role.DefaultTTL.Seconds(),
-			"max_ttl":               role.MaxTTL.Seconds(),
-		}
-		if len(role.Statements.Creation) == 0 {
-			data["creation_statements"] = []string{}
+		db.RLock()
+		defer db.RUnlock()
+
+		if err := db.RevokeUser(ctx, role.Statements, role.StaticAccount.Username); err != nil {
+			b.CloseIfShutdown(db, err)
+			return nil, err
 		}
-		if len(role.Statements.Revocation) == 0 {
-			data["revocation_statements"] = []string{}
+	}
+
+	err = req.Storage.Delete(ctx, databaseStaticRolePath+name)
+	if err != nil {
+		return nil, err
+	}
+
+	if b.credRotationQueue != nil {
```

Comment:
I think you need to hold the read lock to safely do this

Reply:
This has been resolved in a refactoring